### PR TITLE
Making UseExplicitType fix CS0822

### LINKED ISSUE #30516: Code fix for CS0822 (Implicitly-typed variables cannot be constant)
There should be a code fix to use explicit type when this compiler error occurs:
```c#
class C
{
    void M()
    {
        const var v = ""; // CS0822: Implicitly-typed variables cannot be constant
    }
}
```
to
```c#
class C
{
    void M()
    {
        const string v = "";
    }
}
```

### PR DESCRIPTION
**Pull request by Neme12**
Fixes #30516

### CODE REVIEW

File: src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs [hunk 1]
```diff
@@ -21,8 +21,10 @@ namespace Microsoft.CodeAnalysis.CSharp.TypeStyle
     [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseExplicitType), Shared]
     internal class UseExplicitTypeCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs]: i'm definitely not a fan of reusing this fixer for that compiler code.  These fixers are really totally intended to be for the actual cases of the IDE analyzers which validate the users's 'var/explicit' type options against the code.  I think it really muddies the waters and makes things less clear when this is used in such a fashion..--

--

To give some perspective, i've put through nearly a dozen PRs just to detangle cases where we had fixers trying to handle too many cases.  It is usually *much* easier to maintain and understand if we keep the different parts simple and clear as to what their responsibility is.

File: src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs [hunk 2]
```diff
@@ -51,6 +53,13 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             SyntaxNode node, CancellationToken cancellationToken)
         {
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (node is VariableDeclarationSyntax variableDeclaration)
+            {
+                // CS0822 is reported on the declaration, not the type.
+                node = variableDeclaration.Type;
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs]: Do not like this.  We're effectively munging together different diagnostics, then trying to piece together what is going on on the other end.

File: src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs [hunk 3]
```diff
@@ -83,7 +92,11 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             if (parensDesignation is null)
             {
-                var typeSymbol = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded()).ConvertedType;
+                var typeSymbol = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded(), cancellationToken).ConvertedType;
+                if (typeSymbol.IsAnonymousType)
+                {
+                    typeSymbol = semanticModel.Compilation.GetSpecialType(SpecialType.System_Object);
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs]: don't like this either.

**Comment by CyrusNajmabadi** [comment]
It feels like it would be much simpler and clearer if we actually just had a different fixer here.  It would simpy register for the specified diagnostic, get the node, then replace the node with the inferred type.  Very simple and easy without overlapping this other distinct feature.

**Comment by Neme12** [comment]
@CyrusNajmabadi But then if the user encounters this error and has "use explicit type" as their preference, they would get both fixes offered. Do you think that's ok?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi But then if the user encounters this error and has "use explicit type" as their preference, they would get both fixes offered. Do you think that's ok?

I'm actually surprised that "const var" is affected by teh "use explicit type" feature at all.  "const var" isn't even legal code, so it's not really intended to be covered by that feature.  After all, it's not like "use implicit type" will convert a "const int" to "const var".  So i would not expect "use explicit type" to be anything that offers to convert "const var" to "const int".  

If you're adding this feature, i would simply make "use explicit type" not operate at all on the invalid construct "const var".

Thanks!

**Comment by Neme12** [comment]
I'll make a new PR.

File: src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs [hunk 4]
```diff
@@ -60,44 +60,22 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
             }
 
             if (parensDesignation is null)
             {
-                var typeSymbol = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded(), cancellationToken).ConvertedType;
-                if (typeSymbol.IsAnonymousType)
-                {
-                    typeSymbol = semanticModel.Compilation.GetSpecialType(SpecialType.System_Object);
-                }
-
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs]: i would definitely not touch this code if not necessary.

**Pull request closed without merging**